src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_posix.cc

```diff
@@ -98,12 +105,89 @@ class GrpcPolledFdPosix : public GrpcPolledFd {
 
 class GrpcPolledFdFactoryPosix : public GrpcPolledFdFactory {
  public:
+  ~GrpcPolledFdFactoryPosix() {
+    for (auto& fd : owned_fds_) {
+      close(fd);
+    }
+  }
+
   GrpcPolledFd* NewGrpcPolledFdLocked(
       ares_socket_t as, grpc_pollset_set* driver_pollset_set) override {
+    auto insert_result = owned_fds_.insert(as);
+    GPR_ASSERT(insert_result.second);
     return new GrpcPolledFdPosix(as, driver_pollset_set);
   }
 
-  void ConfigureAresChannelLocked(ares_channel /*channel*/) override {}
+  void ConfigureAresChannelLocked(ares_channel channel) override {
+    ares_set_socket_functions(channel, &kSockFuncs, this);
+  }
+
+ private:
+  /// Overridden socket API for c-ares
+  static ares_socket_t Socket(int af, int type, int protocol,
+                              void* /*user_data*/) {
+    int fd = socket(af, type, protocol);
+    /* Because we're using socket API overrides, c-ares won't
+     * perform its typical configuration on the socket. See
+     * https://github.com/c-ares/c-ares/blob/bad62225b7f6b278b92e8e85a255600b629ef517/src/lib/ares_process.c#L1018.
+     * So we copy the default settings that c-ares would
+     * normally apply on posix platforms:
+     *   - non-blocking
+     *   - cloexec flag
+     *   - disable nagle */
+    grpc_error_handle err;
+    err = grpc_set_socket_nonblocking(fd, true);
+    if (!err.ok()) return -1;
```

Comment:
Do you need to close the socket when the setting failed? Same below.

Reply:
Ah, looks like there is also this [ares_set_socket_configure_callback](https://source.corp.google.com/piper///depot/google3/third_party/ares/include/ares.h;bpv=1;bpt=1;rcl=550417489;l=401) which might be used for this purpose? And c-ares will call close socket when it returns -1: https://source.corp.google.com/piper///depot/google3/third_party/ares/src/lib/ares_process.c;l=1154;bpv=0;bpt=1;rcl=550417489

Reply:
Regarding the [ares_set_socket_configure_callback](https://source.corp.google.com/piper///depot/google3/third_party/ares/include/ares.h;bpv=1;bpt=1;rcl=550417489;l=401), I think it may be simpler to close upon failure within here?

Reply:
I would recommend utilizing it since it seems to be specifically designed for this purpose. It would also call our version's of `Close()` (though it does not make any difference). But I wouldn't block on that.

Reply:
done. And I agree now the config fallback does seem make this easier to read